receiver/awsfirehosereceiver/internal/unmarshaler/cwlog/unmarshaler.go

```diff
@@ -37,60 +45,56 @@ func NewUnmarshaler(logger *zap.Logger) *Unmarshaler {
-	md := plog.NewLogs()
-	builders := make(map[resourceAttributes]*resourceLogsBuilder)
-	for recordIndex, compressedRecord := range records {
-		record, err := compression.Unzip(compressedRecord)
-		if err != nil {
-			u.logger.Error("Failed to unzip record",
+func (u Unmarshaler) UnmarshalLogs(compressedRecord []byte) (plog.Logs, error) {
+	r, err := gzip.NewReader(bytes.NewReader(compressedRecord))
+	if err != nil {
+		return plog.Logs{}, fmt.Errorf("failed to decompress record: %w", err)
+	}
+	decoder := json.NewDecoder(r)
+
+	// Multiple logs in each record separated by newline character
+	logs := plog.NewLogs()
+	for datumIndex := 0; ; datumIndex++ {
+		var log cWLog
+		if err := decoder.Decode(&log); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			u.logger.Error(
+				"Unable to unmarshal input",
 				zap.Error(err),
-				zap.Int("record_index", recordIndex),
+				zap.Int("datum_index", datumIndex),
 			)
 			continue
 		}
-		// Multiple logs in each record separated by newline character
-		for datumIndex, datum := range bytes.Split(record, []byte(recordDelimiter)) {
-			if len(datum) > 0 {
-				var log cWLog
-				err := json.Unmarshal(datum, &log)
-				if err != nil {
-					u.logger.Error(
-						"Unable to unmarshal input",
-						zap.Error(err),
-						zap.Int("datum_index", datumIndex),
-						zap.Int("record_index", recordIndex),
-					)
-					continue
-				}
-				if !u.isValid(log) {
-					u.logger.Error(
-						"Invalid log",
-						zap.Int("datum_index", datumIndex),
-						zap.Int("record_index", recordIndex),
-					)
-					continue
-				}
-				attrs := resourceAttributes{
-					owner:     log.Owner,
-					logGroup:  log.LogGroup,
-					logStream: log.LogStream,
-				}
-				lb, ok := builders[attrs]
-				if !ok {
-					lb = newResourceLogsBuilder(md, attrs)
-					builders[attrs] = lb
-				}
-				lb.AddLog(log)
-			}
+		if !u.isValid(log) {
+			u.logger.Error(
+				"Invalid log",
+				zap.Int("datum_index", datumIndex),
+			)
+			continue
 		}
-	}
 
-	if len(builders) == 0 {
-		return plog.NewLogs(), errInvalidRecords
-	}
+		rl := logs.ResourceLogs().AppendEmpty()
+		resourceAttrs := rl.Resource().Attributes()
+		resourceAttrs.PutStr(conventions.AttributeCloudAccountID, log.Owner)
+		resourceAttrs.PutStr(attributeAWSCloudWatchLogGroupName, log.LogGroup)
+		resourceAttrs.PutStr(attributeAWSCloudWatchLogStreamName, log.LogStream)
 
-	return md, nil
+		logRecords := rl.ScopeLogs().AppendEmpty().LogRecords()
+		for _, event := range log.LogEvents {
+			logRecord := logRecords.AppendEmpty()
+			// pcommon.Timestamp is a time specified as UNIX Epoch time in nanoseconds
+			// but timestamp in cloudwatch logs are in milliseconds.
+			logRecord.SetTimestamp(pcommon.Timestamp(event.Timestamp * int64(time.Millisecond)))
+			logRecord.Body().SetStr(event.Message)
+		}
+	}
+	if logs.ResourceLogs().Len() == 0 {
+		return logs, errInvalidRecords
+	}
+	pdatautil.GroupByResourceLogs(logs.ResourceLogs())
```

Comment:
Is there a benchmark to compare the effect of allocating a new resource log entry for each log record and combining them later?  If not, please add one.  This feels like something that could be a significant regression.

Reply:
There's not, I've added one. This way of grouping is indeed much more expensive than what was there before. I've reverted to grouping by the known CloudWatch attributes.

Reply:
Addressed in my other merged PR, resolving.